src/plugins/intel_npu/src/common/include/intel_npu/common/igraph.hpp

```diff
@@ -0,0 +1,117 @@
+#include "intel_npu/utils/zero/zero_wrappers.hpp"
+
+namespace intel_npu {
+
+/**
+ * @struct CompiledNetwork
+ * @brief Custom container for compiled network, used for export
+ * @var CompiledNetwork::data
+ * Pointer to the address of compiled network
+ * @var CompiledNetwork:size
+ * Size of the compiled network
+ * @var CompiledNetwork::ownedStorage
+ * Plugin owned compiled network storage that is required in case of a driver that
+ * doesn't support graph extension 1.7, as in this case plugin must create a copy of the compiled network.
+ * @note It's unsafe to store either data or size outside of the compiled network object as its destructor
+ * would release the owning container
+ */
+
+struct CompiledNetwork {
+    const uint8_t* data;
+    size_t size;
+    CompiledNetwork(const uint8_t* data, size_t size, std::vector<uint8_t> storage)
+        : data(data),
+          size(size),
+          ownedStorage(std::move(storage)) {}
+
+private:
+    std::vector<uint8_t> ownedStorage;
+};
+
+class IGraph : public std::enable_shared_from_this<IGraph> {
+public:
+    IGraph(ze_graph_handle_t handle, NetworkMetadata metadata) : _handle(handle), _metadata(std::move(metadata)) {}
+
+    virtual CompiledNetwork export_blob() const = 0;
+
+    virtual std::vector<ov::ProfilingInfo> process_profiling_output(const std::vector<uint8_t>& profData) const = 0;
+
+    virtual void set_argument_value(uint32_t argi, const void* argv) const = 0;
+
+    virtual void initialize() = 0;
+
+    virtual ~IGraph() = default;
+
+    const NetworkMetadata& get_metadata() const {
+        return _metadata;
+    }
+
+    ze_graph_handle_t get_handle() const {
+        return _handle;
+    }
+
+    void update_network_name(std::string_view name) {
+        _metadata.name = name;
+    }
+
+    inline const std::vector<ArgumentDescriptor>& get_input_descriptors() const {
+        return _input_descriptors;
+    }
+
+    inline const std::vector<ArgumentDescriptor>& get_output_descriptors() const {
+        return _output_descriptors;
+    }
+
+    inline const std::shared_ptr<CommandQueue>& get_command_queue() const {
+        return _command_queue;
+    }
+
+    void setWorkloadType(const ov::WorkloadType workloadType) const {
+        ze_command_queue_workload_type_t zeWorkloadType;
+        switch (workloadType) {
+        case ov::WorkloadType::DEFAULT:
+            zeWorkloadType = ze_command_queue_workload_type_t::ZE_WORKLOAD_TYPE_DEFAULT;
+            break;
+        case ov::WorkloadType::EFFICIENT:
+            zeWorkloadType = ze_command_queue_workload_type_t::ZE_WORKLOAD_TYPE_BACKGROUND;
+            break;
+        default:
+            OPENVINO_THROW("Unknown value for WorkloadType!");
+        }
+
+        _command_queue->setWorkloadType(zeWorkloadType);
+    }
+
+    void mutexLock() {
+        _mutex.lock();
+    }
+
+    void mutexUnlock() {
+        _mutex.unlock();
+    }
+
+protected:
+    ze_graph_handle_t _handle = nullptr;
```

Comment:
As this class is meant as an interface I would highly recommend to remove the members from here.

Especially the handle is only relevant for L0. It might be OK to return zero for CiP and keep the method on interface. Alternative would be to downcast based on some ID or dyn cast.

Reply:
Actually, this would be used even for the CiP case. It will be null only other backends are used.